### PR TITLE
[QDP] [test] add a fidelity test

### LINKED ISSUE #686: [QDP]  [test] add a fidelity test
### Summary
we should add a fidelity test to test our results is same as the results of qiskit circuits.

### PR DESCRIPTION
**Pull request by rich7420**
### Purpose of PR

This tests include: full-stack verification, async pipeline, fidelity metrics,
zero-copy validation, and edge cases (boundaries, stability, memory, threads).

the results:
```
============================= test session starts ==============================
platform linux -- Python 3.12.3, pytest-9.0.1, pluggy-1.6.0 -- /home/rich-wsl/mahout/qdp/qdp-python/.venv/bin/python3
cachedir: .pytest_cache
rootdir: /home/rich-wsl/mahout/qdp/qdp-python
configfile: pyproject.toml
collecting ... collected 12 items

tests/test_high_fidelity.py::test_amplitude_encoding_fidelity_comprehensive[4-16-Small - Sync Path] 
[Test Case] Small - Sync Path (Size: 16)
Fidelity: 1.0000000000000000
PASSED
tests/test_high_fidelity.py::test_amplitude_encoding_fidelity_comprehensive[10-1000-Medium - Padding Logic] 
[Test Case] Medium - Padding Logic (Size: 1000)
Fidelity: 0.9999999999999996
PASSED
tests/test_high_fidelity.py::test_amplitude_encoding_fidelity_comprehensive[18-131072-Boundary - Exact Chunk Size] 
[Test Case] Boundary - Exact Chunk Size (Size: 131072)
Fidelity: 1.0000000000000004
PASSED
tests/test_high_fidelity.py::test_amplitude_encoding_fidelity_comprehensive[18-131073-Boundary - Chunk + 1] 
[Test Case] Boundary - Chunk + 1 (Size: 131073)
Fidelity: 1.0000000000000004
PASSED
tests/test_high_fidelity.py::test_amplitude_encoding_fidelity_comprehensive[18-262144-Boundary - Two Exact Chunks] 
[Test Case] Boundary - Two Exact Chunks (Size: 262144)
Fidelity: 0.9999999999999993
PASSED
tests/test_high_fidelity.py::test_amplitude_encoding_fidelity_comprehensive[20-1000000-Large - Async Pipeline] 
[Test Case] Large - Async Pipeline (Size: 1000000)
Fidelity: 1.0000000000000009
PASSED
tests/test_high_fidelity.py::test_complex_integrity 
Sum of imaginary parts (should be 0): 0.0
PASSED
tests/test_high_fidelity.py::test_numerical_stability_underflow PASSED
tests/test_high_fidelity.py::test_memory_leak_quantitative 
Start GPU Memory: 0 bytes
End GPU Memory:   0 bytes
PASSED
tests/test_high_fidelity.py::test_memory_safety_stress 
Starting memory stress test (1000 iterations)...
Memory stress test passed (no crash).
PASSED
tests/test_high_fidelity.py::test_multithreaded_access 
Starting concurrent stress test with 8 threads...
Multithreaded access check passed.
PASSED
tests/test_high_fidelity.py::test_error_propagation PASSED

============================== 12 passed in 3.33s ==============================
```

### Related Issues or PRs

Closes #686 

### Changes Made

- [ ] Bug fix
- [ ] New feature
- [ ] Refactoring
- [ ] Documentation
- [x] Test
- [ ] CI/CD pipeline
- [ ] Other

### Breaking Changes

- [ ] Yes
- [x] No

### Checklist

- [ ] Added or updated unit tests for all changes
- [ ] Added or updated documentation for all changes
- [x] Successfully built and ran all unit tests or manual tests locally
- [ ] PR title follows "MAHOUT-XXX: Brief Description" format (if related to an issue)
- [ ] Code follows ASF guidelines

### CODE REVIEW

**Comment by rich7420** [comment]
There's a test for ensuring our output.
It ensures async pipeline correctness and the zero-copy path through rust -> CUDA -> pytorch.

File: qdp/qdp-python/tests/test_high_fidelity.py [hunk 1]
```diff
@@ -0,0 +1,235 @@
...
+    num_qubits = 12
+    data_size = 3000  # Non-power-of-2 size
+
+    raw_data = np.random.rand(data_size).astype(np.float64)
+    qtensor = engine.encode(raw_data.tolist(), num_qubits, "amplitude")
+    torch_state = torch.from_dlpack(qtensor)
+
+    imag_error = torch.sum(torch.abs(torch_state.imag)).item()
+    print(f"\nSum of imaginary parts (should be 0): {imag_error}")
+    assert imag_error == 0.0, "State vector contains non-zero imaginary components!"
```
Comment (400Ping) [inline qdp/qdp-python/tests/test_high_fidelity.py]: If the backend ever introduces more floating-point operations, we might start seeing tiny numerical noise (~1e-16) even for real-valued encodings. Maybe we could relax this to something like `assert imag_error <= 1e-14` (or similar), so the test still guards the invariant without being too brittle.

Comment (rich7420) [reply qdp/qdp-python/tests/test_high_fidelity.py]: Cool , I think this vision is good. I'll modify this part.

**Review by ryankert01 — APPROVED** [review]
